src/main/java/edu/ucla/library/iiif/fester/handlers/PostThumbnailsHandler.java

```diff
@@ -0,0 +1,207 @@
+        final Set<FileUpload> csvUploads = aContext.fileUploads();
+        final String festerizeUserAgent = StringUtils.trimTo(request.getHeader("User-Agent"), Constants.EMPTY);
+        final Matcher festerizeUserAgentMatcher = myFUAPattern.matcher(festerizeUserAgent);
+        final String errorMessage;
+
+        // An uploaded CSV is required
+        if (csvUploads.isEmpty()) {
+            errorMessage = LOGGER.getMessage(MessageCodes.MFS_037);
+            returnError(response, HTTP.BAD_REQUEST, errorMessage);
+        } else if (festerizeUserAgentMatcher.matches() &&
+                   !festerizeUserAgentMatcher.group("version").equals(myFesterizeVersion)) {
+            errorMessage = LOGGER.getMessage(MessageCodes.MFS_147, myFesterizeVersion);
+            returnError(response, HTTP.BAD_REQUEST, errorMessage);
+        } else {
+            final FileUpload csvFile = csvUploads.iterator().next();
+            final String filePath = csvFile.uploadedFileName();
+            final String fileName = csvFile.fileName();
+
+            try {
+                final List<String[]> originalLines =
+                    new CSVReader(Files.newBufferedReader(Paths.get(filePath))).readAll();
+                final List<String[]> linesWithThumbs = ThumbnailUtils.addThumbnailColumn(originalLines);
+                final int manifestIndex = Arrays.asList(linesWithThumbs.get(0)).indexOf(CSV.MANIFEST_URL);
+                final String collectionURL = linesWithThumbs.get(1)[manifestIndex];
+                final Future<JsonObject> collection = getManifest(collectionURL);
+                collection.onComplete(result -> {
+                    if (result.failed()) {
+                        returnError(response, HTTP.INTERNAL_SERVER_ERROR, result.cause().getMessage());
+                    } else {
+                        final JsonArray canvases = result.result().getJsonArray(MANIFESTS);
+                        final int canvasCount = canvases.size();
+                        final int canvasIndex;
+                        if (canvasCount <= 3) {
+                            canvasIndex = 0;
+                        } else {
+                            canvasIndex = ThumbnailUtils.pickThumbnailIndex(canvasCount - 1);
+                        }
+                        final String canvasURL = canvases.getJsonObject(canvasIndex).getString(ID_TAG);
+                        final Future<JsonObject> canvas = getManifest(canvasURL);
+                        canvas.onComplete(canvasResult -> {
+                            if (canvasResult.failed()) {
+                                returnError(response, HTTP.INTERNAL_SERVER_ERROR, canvasResult.cause().getMessage());
+                            } else {
+                                final JsonObject thumbCanvas = canvasResult.result();
+                                final String thumbURL = thumbCanvas.getJsonArray("sequences").getJsonObject(0)
```

Comment:
(Just noting that this expression for getting the thumbnail URL is also v2-specific.)

Reply:
Just to elaborate a bit on [Kevin's comment about wanting to use base URLs](https://github.com/UCLALibrary/fester/pull/162#pullrequestreview-582618047), this is basically the URL of the IIIF Image API service for the image. To retrieve:
- v2: `canvas["images"][0]["resource"]["service"]["@id"]`
- v3: `canvas["items"][0]["items"][0]["body"]["service"]["id"]` (note no `@` for v3 IDs)

API version of the input data can be determined as explained [here](https://github.com/UCLALibrary/fester/pull/162#discussion_r569829714)